vault/client.go

```diff
@@ -1,84 +1,167 @@
 package vault
 
 import (
+	"context"
 	"encoding/base64"
 	"encoding/json"
 	"fmt"
 	"io/ioutil"
 	"log"
+	"sync"
+	"time"
 
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/hashicorp/vault-lambda-extension/config"
 	"github.com/hashicorp/vault/api"
 )
 
+// Client holds api.Client and handles state required to renew tokens and re-auth as required.
+type Client struct {
+	mtx sync.Mutex
+
+	VaultClient *api.Client
+	VaultConfig *api.Config
+
+	logger     *log.Logger
+	stsSvc     *sts.STS
+	authConfig config.AuthConfig
+
+	// Token refresh/renew data.
+	tokenExpiry    time.Time
+	tokenTTL       time.Duration
+	tokenRenewable bool
+}
+
 // NewClient uses the AWS IAM auth method configured in a Vault cluster to
 // authenticate the execution role and create a Vault API client.
-func NewClient(logger *log.Logger, vaultAuthRole, vaultAuthProvider string, vaultIAMServerID string) (*api.Client, *api.Config, error) {
-	config := api.DefaultConfig()
-	if config.Error != nil {
-		return nil, nil, fmt.Errorf("error making default vault config for extension: %w", config.Error)
-	}
-	vaultClient, err := api.NewClient(config)
+func NewClient(logger *log.Logger, vaultConfig *api.Config, authConfig config.AuthConfig) (*Client, error) {
+	vaultClient, err := api.NewClient(vaultConfig)
 	if err != nil {
-		return nil, nil, fmt.Errorf("error making extension: %w", err)
+		return nil, fmt.Errorf("error making extension: %w", err)
 	}
 
 	ses := session.Must(session.NewSession())
 	stsSvc := sts.New(ses)
 
+	client := &Client{
+		VaultClient: vaultClient,
+		VaultConfig: vaultConfig,
+
+		logger:     logger,
+		stsSvc:     stsSvc,
+		authConfig: authConfig,
+	}
+
+	return client, nil
+}
+
+// Token synchronously renews/re-auths as required and returns a Vault token.
+func (c *Client) Token(ctx context.Context) (string, error) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
```

Comment:
Minor nit: every request handled by the `proxyHandler` is going to capture a full lock here. It might not be a problem though. If it turns out to be a problem we'd likely want to do use a read/write mutex and "upgrade" the lock here (it's not really upgrading but you get my point). 

No action required here, just posting

Reply:
After more reflection, I'm not sure this is really something worth looking into at all

Reply:
I went through the same thought process as you 🙂 I actually started writing this with a `RWMutex` and then decided it was probably too early to make that optimisation without some data. Assuming a reasonable token TTL and a scenario where traffic is high enough that we care about lock contention, most calls to this function will already have a valid token that doesn't need renewing, so I think it will probably be quick enough.